src/Client/HedgedConnections.h

```diff
@@ -0,0 +1,155 @@
+#pragma once
+#if defined(OS_LINUX)
+
+#include <functional>
+#include <queue>
+#include <Client/HedgedConnectionsFactory.h>
+#include <Client/IConnections.h>
+
+namespace DB
+{
+
+/** To receive data from multiple replicas (connections) from one shard asynchronously,
+  * The principe of Hedged Connections is used to reduce tail latency:
+  * (if we don't receive data from replica for a long time, we try to get new replica
+  * and send query to it, without cancelling working with previous replica). This class
+  * supports all functionality that MultipleConnections has.
+  */
+class HedgedConnections : public IConnections
+{
+public:
+    struct ReplicaState
+    {
+        Connection * connection = nullptr;
+        std::unordered_map<int, ConnectionTimeoutDescriptorPtr> active_timeouts;
+    };
+
+    struct ReplicaLocation
+    {
+        size_t offset;
+        size_t index;
+    };
+
+    struct OffsetState
+    {
+        std::vector<ReplicaState> replicas;
+        size_t active_connection_count;
+        bool first_packet_of_data_received;
+    };
+
+    HedgedConnections(const ConnectionPoolWithFailoverPtr & pool_,
+                      const Settings & settings_,
+                      const ConnectionTimeouts & timeouts_,
+                      const ThrottlerPtr & throttler,
+                      PoolMode pool_mode,
+                      std::shared_ptr<QualifiedTableName> table_to_check_ = nullptr);
+
+    void sendScalarsData(Scalars & data) override;
+
+    void sendExternalTablesData(std::vector<ExternalTablesData> & data) override;
+
+    void sendQuery(
+        const ConnectionTimeouts & timeouts,
+        const String & query,
+        const String & query_id,
+        UInt64 stage,
+        const ClientInfo & client_info,
+        bool with_pending_data) override;
+
+    Packet receivePacket() override;
+
+    Packet receivePacketUnlocked(AsyncCallback async_callback) override;
+
+    void disconnect() override;
+
+    void sendCancel() override;
+
+    void sendIgnoredPartUUIDs(const std::vector<UUID> & uuids) override;
+
+    Packet drain() override;
+
+    std::string dumpAddresses() const override;
+
+    size_t size() const override { return offset_states.size(); }
+
+    bool hasActiveConnections() const override { return active_connection_count > 0; }
+
+private:
+    /// We will save actions with replicas in pipeline to perform them on the new replicas.
+    class Pipeline
```

Comment:
Also here we need a comment why actually we need a pipeline.
The reason is that we want repeat every used interface method before `receivePacket`, like `sendScalarsData` or `sendExternalTablesData`, for every new replica which was added in case of hedging.

BTW, I think that this solution is easy, but clumsy a little bit, cause we actually call interface methods in the same order all the time. We can just probably init connection with all needed data and call `receivePacketUnlocked` only.
It is difficult for now because we heed to be able to witch to prev implementation, and so have `IConnections` iface.

Probably, later it would be better to refactor this part somehow. E.g. move some of logic from `RemoteQueryExecutor` to connection or create another wrapper for not hedged case.